README.rst
Spiny
=====

Spiny Norman is a Hedgehog.

It's also a package that will run your Python tests under multiple versions
of Python.

"Hey hang on!" I hear you say, "There's already Tox!" And you are right.
Spiny is my attempt to look into the problem space tox covers to make
something that is less complex, and fixes a couple of problems with Tox.

Most notably, with Spiny, Python does not have to be on the path, it can
install versions of Python other than the one your installed version of
virtualenv happens to support, and it does not always call setup.py.
It also will run the tests in parallel.

Major feature: For the common use case it needs no configuration.

There is no guarantee that Spiny's features will not end up in Tox.


Command line parameters
-----------------------

You run Spiny from the root of the Python project that you want to test,
ie the direectory that has the ``setup.py`` file.

The command line parameters are:

  usage: spiny [-h] [--version] [-c <filename>] [-e <environments>] [-v] [-q]
               [<configvar> [<configvar> ...]]

  Run tests under several Python versions.

  positional arguments:
    <configvar>           Override a config variable by "section:variable=value"
                          Example: "spiny:venv-dir=.venv"

  optional arguments:
    -h, --help            Show this help message and exit.
    --version             Show the version and exit.
    -c <filename>, --config <filename>
                          The config file to use. Defaults "to spiny.cfg".
    -e <environments>, --envlist <environments>
                          A list of environments to run, separated by commas.
    -v, --verbose         Increases the output, -vv increases it even more.
    -q, --quiet           Reduces output to only the run summary, -qq removes
                          also that.


Version support
---------------

Spiny can be run under Python 2.6, 2.7, 3.3 and 3.4. It can also be run under
PyPy, PyPy3 and Jython.

It can run tests under a much wider range of Python versions, this has been tested
with Python 2.4, 2.5, 3.1 and 3.2 in addition to the above Python versions.

IronPython is supported in theory, but I can't get virtualenv working with
IronPython. Other Python implementations are not tested. If one doesn't work,
you are welcome to raise an issue and I can look into it.


Configuration files
-------------------

Spiny does not typically need a configuration file. It will instead look at
your ``setup.py`` and find out what Python versions you module supports, and
run the tests with those versions.

You declare version support with classifiers, like this::

  setup(
      ...
      classifiers=[
          ...
          "Programming Language :: Python",
          "Programming Language :: Python :: 2.6",
          "Programming Language :: Python :: 2.7",
          "Programming Language :: Python :: 3.3",
          "Programming Language :: Python :: 3.4",
          "Programming Language :: Python :: Implementation :: PyPy",
          ...
          ],
      ...
  )

With these classifiers, Spiny will run the tests under Python 2.6, 2.7, 3.3,
3.4 and also under PyPy.

You can configure Spiny explicitly on a configuration basis. You can add a
``spiny.cfg`` located in the project root, or you can add the configuration
to the projects ``setup.cfg``.

You can also have a personal configuration file in ``~/.config/spiny.cfg``
whose configuration options will be valid for all projects.

Under the ``[spiny]`` section the
following configuration options are supported:

  * **environments**: A whitespace separated list of Python and Python versions.

  * **venv-dir**: The name of the directory to install virtualenvs in.
    Defaults to ``.venv``.

  * **max-processes**: The maxiumum of concurrent processes to run tests with.
    Defaults to the number of CPU's you have.

  * **test-commands**: The commands used to run the tests. You can have
    several lines of commands. Defaults to ``{envpython} setup.py test``. There
    are a few variables that you can use in the commands that will be replaced:

    * ``{envpython}`` will be replaced with the full path to the Python
      executable in the virtualenv

    * ``{basepython}`` will be replaced with the full path to the Python
      executable the virtualenv is created from

    * ``{envdir}`` will be replaced with the full path to the virtualenv
      directory.

    * ``{projectdir}`` will be replaced with the full path to the directory
      of the Python project (ie, the current directory)

  * **setup-commands**: The commands used to create the virtualenv. The default
    for this varies, but it boils down to ``{envpython} -m virtualenv {envdir}``.

  * **use-setup-py**: If requirements data from ``setup.py`` should be used to
    gather requirements. This means ``setup.py`` needs to exist, and be
    executable without side-effects. Defaults to ``true``.

  * **use-requirements-txt**: If requirements data from ``requiremenets.txt``
    should be used to gather requirements. Defaults to ``true``.

  * **changedir**: A directory to change to before running the tests.
    Variables from test-commands are usable.


Example::

  [spiny]
  environments = python2.7
                 python3.4
                 pypy2.4

  venv-dir = .venv

  test-commands = {envpython} something.py magic
                  {envpython} setup.py test

  max-processes = 3

There is also a ``[pythons]`` section, which defines up the paths to the various
executables, per environment::

  [pythons]
  python2.6 = /pythons/python26/bin/python
  python2.7 = /pythons/python27/bin/python
  python3.3 = /pythons/python33/bin/python3
  python3.4 = /pythons/python34/bin/python3

If you don't configure this, the executables that are on the PATH will be used.

This doesn't make much sense to have in the projects ``spiny.cfg``,
as each person who runs the tests are likely to have differing Python installs.
However, this does make a lot of sense to have in the personal configuration file.

You can add the ``[pythons]`` section in your projects ``spiny.cfg``, but the
usecase for that is very limited. Possibly if you are using custom Pythons in
your project.

``max-processes`` also is reasonable in your personal file, if you for
example have very long-running tests, and you want to keep a CPU free, for
example for browing the web while the tests run. It also makes sense in a
project file if your tests use a lot of memory, to avoid running out of
memory.

``environments`` and ``test-commands`` only make sense per configuration and
not in the personal file. However, no checks for this are done, so you can
add them there if you want to, but the results are unlikely to be practical.


Todo
----

Things that needs doing:

  * Make the tests run under itself.

  * Figure out how to run coverage on things run by subprocesses.

  * Add commands per environment, enabling things like a pep8 environment
    that checks for pep8 compliance, etc.

  * Windows support. Maybe.


docs/CHANGES.txt
CHANGES
=======

0.7 (unreleased)
----------------

- Fixed a bug on Python pre-release versions.


0.6 (2017-04-12)
----------------

- Using /dev/null as stdin means the tests doesn't hang, but errors out.

- No redirection of test output (or input) when only using one process.

- It's not a library, so supporting Python 2 isn't needed.


0.5 (2016-02-26)
----------------

- CTRL-C will now kill the program.

- Renamed {envpath} to {envdir}, {python} to {envpython} and {project_dir} to
  {projectdir}. Also added {basepython} to the Python version used to install
  the virtual env.

- Various changes to make custom commands more stable and flexible.

- Added a changedir option to the config.

- You can now run tests for Pythons that does not support any version
  of virtualenv, such as Python 2.3 and earlier.

- Gives a nice error message and exits with -1 if it doesn't know what
  versions to use.

- Dropped support for running Spiny with Python 2.6.

- If using a system Python 3 the virtualenv will no longer install system
  Python 2.

- Spiny now supports having different requirements in setup.py for different
  versions of Python.

- Not finding an executable for a specific environment is no longer a fatal
  error, so you don't have to have all the supported versions installed.


0.4 (2014-11-01)
----------------

- Added support for requirements.txt.

- Added parameters to skip executing setup.py.

- Added a setup-commands parameter to override how the virtualenv is made.

- The -e parameter now actually works.

- The information about the python execs is now cached. This speeds up
  startup, a lot, fetching the Python environments goes from around 5
  seconds to less than a tenth.

- A profile of each virtualenv is now saved in the virtualenv, containing
  it's name, the full path of the Python exe, and the installed requirements.
  A virtualenv will now only be updated if this has changed since the last run.
  This also speeds up tests runs a lot in the typical case.


0.3 (2014-10-30)
----------------

- Now runs under Python 3, PyPy, PyPy3, Jython.

- All output is now through logging, which enables you to select verbosity.

- Added --version, --verbose and --quiet options.

- Now use subprocess32 under Python 2.

- Will now run tests in parallell with multiprocessing. A 'max-processes'
  option in the [spiny] section will allow you to cap the processes if
  you don't want to max your computer. Otherwise it will use as many processes
  as multiprocessing will report cpu's.


0.2 (2014-10-27)
----------------

- Now prints a summary in the end.

- Regard more virtualenv tests as failures.

- Return 1 if not all tests succeeded.

- Fall back to setup.cfg is there is no spiny config file.

- Renamed spiny.conf to spiny.cfg for consistency.

- Made the configuration file optional.

- Spiny will now grep setup.py for Python version classifiers if no
  environment configuration is found.

- If both Python X and Python X.Y is given as environments, it will
  skip the tests for Python X, seeing as it is fulfilled by Python X.Y.

- The version specified in the [pythons] section now have preference.

- Setting a config option from the command line now works even if the section
  doesn't exist.

- Spiny will now attempt to extract the test requirements and install them
  into the virtualenv before running the tests, to avoid polluting the
  project directory with .egg directories.


0.1.1 (2014-10-18)
------------------

- Initial release


tests/utils.py
import os
import shutil
import spiny
import tempfile

try:
    from ConfigParser import ConfigParser
except ImportError:
    from configparser import ConfigParser


class TestEnvironment(object):

    def __init__(self, env_list):
        self.env_list = env_list
        # By passing in an empty conf, we get the Pythons that are in the path only:
        conf = make_conf()
        conf.set('spiny', 'environments', ' '.join(env_list))
        self.pythons = spiny.environment.get_pythons(conf)

        for env in self.env_list:
            if env not in self.pythons:
                raise EnvironmentError("You must have %s installed and "
                                       "on the path to run the tests." % env)

    def __enter__(self):
        self.test_dir = tempfile.mkdtemp()
        try:
            self.old_path = os.environ['PATH']
            self.old_home = os.environ['HOME']
            os.environ['PATH'] = self.test_dir
            os.environ['HOME'] = self.test_dir
            for env in self.env_list:
                path = self.pythons[env]['path']
                os.symlink(path, os.path.join(self.test_dir, env))
            return self
        except Exception:
            # If there is an error here, __exit__ is never called,
            # so we need to cleanup.
            self._cleanup()

    def __exit__(self, exc_type, exc_value, traceback):
        self._cleanup()

    def _cleanup(self):
        os.environ['PATH'] = self.old_path
        os.environ['HOME'] = self.old_home
        shutil.rmtree(self.test_dir)


def make_conf():
    conf = ConfigParser()
    conf.add_section('spiny')
    conf.add_section('pythons')
    return conf


tests/test_main.py
import os
import shutil
import tempfile
import subprocess
import sys
import unittest

import spiny.main


class TestMainBase(unittest.TestCase):

    def setUp(self):
        self.test_dir = tempfile.mkdtemp()
        self.run_dir = os.path.abspath(os.curdir)
        self.pkg_dir = os.path.join(self.test_dir, 'dinsdale')
        self.use_config = os.path.join(self.pkg_dir, os.path.split(self.config_file)[-1])

        shutil.copytree('tests/package', self.pkg_dir)
        shutil.copy(self.config_file, self.pkg_dir)
        os.chdir(self.pkg_dir)
        spiny.main.setup_logging(0, 2)

    def tearDown(self):
        shutil.rmtree(self.test_dir)
        os.chdir(self.run_dir)


class TestMinimal(TestMainBase):
    config_file = 'tests/configs/minimal.conf'

    def test_minimal(self):
        venv_dir = os.path.join(self.test_dir, '.venv')
        sys.argv = [sys.executable, '-qq', '-c', self.use_config,
                    'spiny:venv-dir=%s' % venv_dir]
        spiny.main.main()
        self.assertTrue(os.path.isdir(venv_dir),
                        "The .venv directory was not created")
        self.assertListEqual(['python2.7'], os.listdir(venv_dir))


class TestDual(TestMainBase):
    config_file = 'tests/configs/dual.conf'

    def test_minimal(self):
        venv_dir = os.path.join(self.test_dir, '.venv')
        sys.argv = [sys.executable, '-qq', '-c', self.use_config,
                    'spiny:venv-dir=%s' % venv_dir]
        spiny.main.main()
        self.assertTrue(os.path.isdir(venv_dir),
                        "The .venv directory was not created")
        self.assertListEqual(['python2', 'python3'], sorted(os.listdir(venv_dir)))


tests/package/setup.py
from setuptools import setup, find_packages

setup(
    name='spiny test package',
    version='1.0',
    description='''This is the package whose tests the spiny tests test.''',
    packages=find_packages(),
    author='Lennart Regebro',
    author_email='regebro@gmail.com',
    url="https://github.com/regebro/spiny/",
    license='MIT',
    install_requires=[],
    test_suite='tests'
)


spiny/projectdata.py
# Extracts information from a project that has a distutils setup.py file.
import os
import sys
import string
import logging


def parse_version(version):
    vmap = {'a': 'alpha', 'b': 'beta', 'c': 'candidate'}
    res = []
    version = list(reversed(version))
    v = ''
    while version:
        x = version.pop()
        if x in string.digits:
            v += x
            continue
        else:
            res.append(int(v))
            if v in vmap:
                res.append(vmap[v])
            elif v == ' ':
                break  # Nothing more after the space

    if len(res) == '3':
        res.extend(['final', 0])

    return tuple(res)


class FakeContext(object):

    def __init__(self, path):
        self._path = path

    def __enter__(self):
        self._old_path = os.path.abspath(os.curdir)
        if self._old_path in sys.path:
            sys.path.remove(self._old_path)
        os.chdir(self._path)

        if self._path not in sys.path:
            sys.path.insert(0, self._path)
            self._path_appended = True
        else:
            self._path_appended = False

    def __exit__(self, exc_type, exc_val, exc_tb):
        if self._path_appended:
            sys.path.remove(self._path)
        sys.path.append(self._old_path)

        os.chdir(self._old_path)


class SetupMonkey(object):

    def __init__(self, python_version=None):
        self.python_version = python_version
        self._old_python_version = None

    def distutils_setup_replacement(self, **kw):
        self._kw = kw
        self._kw['_setuptools'] = False

    def setuptools_setup_replacement(self, **kw):
        self._kw = kw
        self._kw['_setuptools'] = True

    def get_data(self):
        return self._kw

    def __enter__(self):
        import distutils.core
        self._distutils_setup = distutils.core.setup
        distutils.core.setup = self.distutils_setup_replacement

        try:
            import setuptools
            self._setuptools_setup = setuptools.setup
            setuptools.setup = self.setuptools_setup_replacement
        except ImportError:
            self._setuptools_setup = None

        self._kw = {}

        if self.python_version is not None:
            # If setup.py checks the Python version, we have to trick it
            # that we are running with the version that we'll run the
            # tests with, and not the version we really are using.
            self._old_python_version = sys.version, sys.version_info
            sys.version = self.python_version
            sys.version_info = parse_version(self.python_version)

        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        import distutils.core
        distutils.core.setup = self._distutils_setup
        if self._setuptools_setup is not None:
            import setuptools
            setuptools.setup = self._setuptools_setup

        if self._old_python_version is not None:
            sys.version, sys.version_info = self._old_python_version


def _specified_versions(data):
    classifiers = data.get('classifiers', [])
    for classifier in classifiers:
        parts = [p.strip() for p in classifier.split('::')]
        if parts[0] == 'Programming Language' and parts[1] == 'Python':
            if len(parts) == 2:
                # Specified Python, but no version.
                continue
            version = parts[2]
            try:
                int(version)
                # This is just specifying 2 or 3, not which version
                continue
            except ValueError:
                pass
            try:
                float(version)
                # This version is good!
                yield version
            except ValueError:
                # Not a proper Python version
                continue


def get_data(path, python_version=None):
    """
    Returns data from a package directory.
    'path' should be an absolute path.
    """
    # Run the imported setup to get the metadata.
    with FakeContext(path):
        with SetupMonkey(python_version) as sm:
            try:
                import setup
                metadata = sm.get_data()

                if not metadata:
                    # This may be a module, like twisted, that only runs setup()
                    # when setup.py is called as the main script. In that case it
                    # often has a main() script to call instead. Try that.
                    try:
                        setup.main()
                    except TypeError:  # OK, so it's twisted.
                        try:
                            setup.main([])
                        except TypeError:
                            pass  # OK, not twisted, then.
                    except AttributeError:
                        pass  # No, no main.

                    metadata = sm.get_data()
                del sys.modules['setup']
            except ImportError as e:
                # Either there is no setup py, or it's broken.
                logging.exception(e)
                metadata = {}

        # No data found
        if not metadata:
            return {}

    return metadata

if __name__ == '__main__':
    import json
    import sys
    path = sys.argv[1]
    print(json.dumps(get_data(path)))


spiny/main.py
import argparse
import logging
import multiprocessing
import os
import os.path
import pkg_resources
import signal
import sys

if sys.version_info < (3,):
    import subprocess32 as subprocess
    from ConfigParser import ConfigParser
else:
    import subprocess
    from configparser import ConfigParser

if sys.platform == 'win32':
    null = 'nul'
else:
    null = '/dev/null'

from spiny import environment, projectdata

__version__ = pkg_resources.require("spiny")[0].version

logger = logging.getLogger('spiny')


class Filter(object):
    """Only log messages that are non-empty."""

    def filter(self, record):
        return len(record.msg)

class LevelFormatter(logging.Formatter):
    """Formatter with a format per level"""
    def __init__(self, fmt=None, datefmt=None, level_formats=None):
        super(LevelFormatter, self).__init__(fmt, datefmt)
        if level_formats is None:
            level_formats = {}
        self._level_formats = level_formats
        self._default_format = fmt

    def format(self, record):
        self._fmt = self._level_formats.get(record.levelno, self._default_format)
        return super(LevelFormatter, self).format(record)


def setup_logging(verbose, quiet):

    verbosity = 2
    if verbose is not None:
        verbosity += verbose
    if quiet is not None:
        verbosity -= quiet

    verbosity = max((0, min(4, verbosity)))
    level = 50-(10*verbosity)

    # The levels can be:
    # 50: Only used for "Oups I failed" messages. -qq
    # 40: Quiet level. Should be used only by the final status message. -q
    # 30: Normal level. "I'm doing this now". Unexpected stderr outputs.
    # 20: Verbose level. All outputs, except when probing. -v
    # 10: Debug level. All outputs, detailed messages. -vv
    # As you notice, these correspone to the logging modules levels
    # CRITICAL, ERROR, WARNING, INFO and DEBUG, but those names
    # don't fit well here. Also, we want different formatting for
    # different levels. So:
    handler = logging.StreamHandler()
    handler.addFilter(Filter())
    formats = {30: '%(message)s', 40: '%(message)s'}
    handler.setFormatter(LevelFormatter('%(message)s',
                                        level_formats=formats))
    logger.setLevel(level)
    logger.addHandler(handler)


def run_all_tests(config):
    """Run a list of commands in each virtualenv"""

    # Get the location of environments.
    if config.has_option('spiny', 'venv-dir'):
        venv_dir = config.get('spiny', 'venv-dir')
    else:
        venv_dir = '.venv'
    venv_dir = os.path.abspath(venv_dir)

    # Get the list of environments to be used:
    pythons = environment.get_pythons(config)
    envnames = environment.get_environments(config)
    if not envnames:
        print("You must specify which Python environments to run tests under, "
              "either in setup.py or with the --envlist argument.")
        sys.exit(1)

    # Get the setup commands:
    if config.has_option('spiny', 'setup-commands'):
        setup_commands = filter(None, config.get('spiny', 'setup-commands').splitlines())
    else:
        setup_commands = None

    # Get the test commands:
    if config.has_option('spiny', 'test-commands'):
        test_commands = filter(None, config.get('spiny', 'test-commands').splitlines())
    else:
        test_commands = ['{envpython} setup.py test']

    if config.has_option('spiny', 'max-processes'):
        max_proc = int(config.get('spiny', 'max-processes'))
    else:
        max_proc = None

    # Get requirements from requirements.txt.
    requirements = []
    if not (config.has_option('spiny', 'use-requirements-txt') and
            config.get('spiny', 'use-requirements-txt').lower() in
            ['false', 'off', '0', 'no']):
        if os.path.isfile('requirements.txt'):
            with open('requirements.txt', 'rt') as reqtxt:
                requirements.extend(reqtxt.readlines())

    if (config.has_option('spiny', 'use-setup-py') and
        config.get('spiny', 'use-setup-py').lower() in
        ['false', 'off', '0', 'no']):
        # Use of setup.py is disabled.
        use_setup = False
    else:
        use_setup = True

    if config.has_option('spiny', 'changedir'):
        curdir = config.get('spiny', 'changedir')
    else:
        curdir = None

    if not os.path.exists(venv_dir):
        os.mkdir(venv_dir)

    projectdir = os.path.abspath(os.path.curdir)

    executes = []
    skips = []
    argslist = []
    for envname in envnames:
        if envname in pythons:
            executes.append(envname)

            # Get requirements from setup.py
            reqs = requirements[:]
            if use_setup:
                project_data = projectdata.get_data('.', pythons[envname]['version'])
                reqs.extend(project_data.get('install_requires', []))
                reqs.extend(project_data.get('setup_requires', []))
                reqs.extend(project_data.get('tests_require', []))
                reqs.extend(project_data.get('extras_require', {}).get('tests', []))
                dependency_links = project_data.get('dependency_links', [])
            else:
                # Use of setup.py is disabled.
                dependency_links = []

            arguments = (envname,
                         pythons[envname],
                         venv_dir,
                         setup_commands,
                         test_commands,
                         reqs,
                         dependency_links,
                         projectdir,
                         curdir)
            argslist.append(arguments)
        else:
            skips.append(envname)

    cpus = min(multiprocessing.cpu_count(), len(executes))
    if max_proc:
        cpus = min(cpus, max_proc)

    all_reqs = [args[5] for args in argslist]
    for req in all_reqs:
        if req != all_reqs[0]:
            # There are different requirements for different versions.
            # Then we can't run the tests in parallell.
            cpus = 1
            logger.log(30, "Version dependent requirements detected, "
                           "not using parallelism.")
            break

    logger.log(20, "Using %s parallel processes" % cpus)
    argslist = [args + (cpus > 1,) for args in argslist]
    pool = multiprocessing.Pool(processes=cpus)
    results = pool.map(run_tests, argslist)
    results = dict(zip(executes, results))
    for envname in skips:
        results[envname] = 'Error: Skipped %s' % envname

    return results


def run_tests(args):
    try:
        (envname, envdict, venv_dir, setup_commands, test_commands,
         requirements, dependency_links, projectdir, curdir, parallel) = args

        if parallel:
            stdout = stderr = subprocess.PIPE
        else:
            # Don't redirect if only one process.
            stdout = stderr = None
        exepath = envdict['path']  # Actual Python exe
        if envdict['virtualenv'] == 'unsupported':
            # Python 2.3 or earlier (or otherwise)
            python = envdict['path']
            envdir = os.path.dirname(os.path.dirname(python))
        else:
            envdir = os.path.join(venv_dir, envname)  # virtualenv dir
            python = os.path.join(envdir, 'bin', envdict['execname'])  # Virtualenv python

        env_parameters = {
            'basepython': exepath,
            'envdir': envdir,
            'envpython': python,
            'projectdir': projectdir,
        }

        # Expand the current directory
        if curdir is not None:
            curdir = curdir.format(**env_parameters)
        else:
            curdir = projectdir

        # Create a "profile"" of this virtualenv, include name, the python exe and requirements.
        venv_profile = '\n'.join([envname, exepath, '\n'.join(sorted(requirements))])
        # Check if there is an existing venv, and in that case read in it's profile:
        profile_path = os.path.join(envdir, '.spiny-profile')
        if os.path.exists(profile_path):
            with open(profile_path, 'rb') as profile:
                installed_profile = profile.read()
        else:
            installed_profile = ''

        if venv_profile != installed_profile:
            # We need to install the virtualenv or update the requirements.

            if not setup_commands:
                if envdict['virtualenv'] == 'internal':
                    # Internal means use the virtualenv for the relevant Python
                    venvexe = exepath
                elif envdict['virtualenv'] == 'external':
                    # External means use the virtualenv for the current Python
                    venvexe = sys.executable
                if envdict['virtualenv'] == 'unsupported':
                    # No virtualenv
                    setup_commands = [[]]
                else:
                    setup_commands = [[venvexe, '-m', 'virtualenv', '-v',
                                       '-p', exepath, envdir]]

            else:
                setup_commands = [command.format(**env_parameters).split() for command in setup_commands]

            logger.log(30, 'Install/update virtualenv for %s' % envname)
            for command in setup_commands:

                # Switch to curdir, if it exists.
                if curdir is not None and os.path.isdir(curdir):
                    os.chdir(curdir)

                logger.log(10, 'Using command: %s' % ' '.join(command))
                with subprocess.Popen(command,
                                      stdout=stdout,
                                      stderr=stderr) as process:
                    process.wait()
                    if parallel:
                        logger.log(30, process.stderr.read())
                        logger.log(20, process.stdout.read())
                    if process.returncode != 0:
                        # This failed somehow
                        msg = "Installing/updating virtualenv for %s failed!" % envname
                        logger.log(30, msg)
                        return msg

            if requirements:
                # Install dependencies:
                pip_path = os.path.join(envdir, 'bin', 'pip')
                parameters = '-f '.join(dependency_links).split()
                parameters.append('-q')
                if envdict['python'] == 'Python' and envdict['version'] < '2.6':
                    # Using 2.5 or worse means no SSL.
                    parameters.append('--insecure')

                command = [pip_path] + parameters + ['install'] + requirements

                logger.log(10, 'Install dependencies with command: %s' % ' '.join(command))
                with subprocess.Popen(command,
                                      stdout=stdout,
                                      stderr=stderr) as process:
                    process.wait()
                    if parallel:
                        logger.log(30, process.stderr.read())  # Log stderr only if verbose output.
                    if process.returncode != 0:
                        # This failed somehow.
                        msg = "Installing/updating dependencies for %s failed!" % envname
                        logger.log(30, msg)
                        if parallel:
                            # pip has the error on stdout. Log it on normal level.
                            logger.log(30, process.stdout.read())
                        return msg
                    elif parallel:
                        # Log successful stdout only if output level is verbse.
                        logger.log(20, process.stdout.read())

            # Save the venv information:
            with open(profile_path, 'wt') as profile:
                profile.write(venv_profile)

        # Run tests:
        logger.log(30, 'Running tests for %s' % envname)

        for command in test_commands:
            command = command.strip().format(**env_parameters)
            logger.log(10, 'Using command: %s' % command)
            with open(null) as nullfile:
                if parallel:
                    stdin = nullfile
                else:
                    stdin = None  # Don't redirect if only one process.
                with subprocess.Popen(command.split(),
                                      stdout=stdout,
                                      stderr=stderr,
                                      stdin=stdin) as process:
                    process.wait()
                    if parallel:
                        # Display the outputs
                        logger.log(30, process.stderr.read().decode())
                        logger.log(30, process.stdout.read().decode())
                    if process.returncode != 0:
                        msg = "Tests failed for %s!" % envname
                        return msg

        return None

    except KeyboardInterrupt:
        return "Tests interrupted by CTRL-C"


def main():
    parser = argparse.ArgumentParser(
        description='Run tests under several Python versions.',
        add_help=False
    )

    parser.add_argument(
        '-h',
        '--help',
        action='help',
        help='Show this help message and exit.')

    parser.add_argument(
        '--version',
        action='version',
        version=__version__,
        help='Show the version and exit.')

    parser.add_argument(
        '-c',
        '--config',
        action='store',
        default='spiny.cfg',
        metavar='<filename>',
        type=str,
        help='The config file to use. Defaults "to spiny.cfg".')

    parser.add_argument(
        '-e',
        '--envlist',
        action='store',
        metavar='<environments>',
        type=str,
        help='A list of environments to run, separated by commas.')

    parser.add_argument(
        '-v',
        '--verbose',
        action='count',
        help='Increases the output, -vv increases it even more.')

    parser.add_argument(
        '-q',
        '--quiet',
        action='count',
        help='Reduces output to only the run summary, -qq removes also that.')

    parser.add_argument(
        'configvar',
        action='store',
        type=str,
        nargs='*',
        metavar='<configvar>',
        help='Override a config variable by "section:variable=value" '
             'Example: "spiny:venv-dir=.venv"')

    args = parser.parse_args()

    setup_logging(args.verbose, args.quiet)

    if args.envlist:
        args.configvar.append('spiny:environments=' + args.envlist.replace(',', ' '))
    return run(args.config, args.configvar)


def run(config_file, overrides):
    # Parse the config files
    if 'HOME' in os.environ:
        home = os.environ['HOME']
    else:
        home = '~'

    # User settings
    settings_file = os.path.join(home, '.config', 'spiny.cfg')

    config = ConfigParser()
    config.read([settings_file, config_file, 'setup.cfg'])

    for override in overrides:
        if ':' not in override or '=' not in override:
            raise ValueError('%s is not a valid config variable. '
                             'It should be "section:variable=value"' % override)
        section, rest = override.split(':', 1)
        option, value = rest.split('=', 1)
        if not config.has_section(section):
            config.add_section(section)
        config.set(section.strip(), option.strip(), value.strip())

    results = run_all_tests(config)

    # Done
    for env in sorted(results):
        if results.get(env):
            logger.log(40, "ERROR: " + results[env])
        else:
            logger.log(40, "       Running tests under %s suceeded." % env)

    return 1 if any(results.values()) else 0


if __name__ == '__main__':
    main()